[css-pseudo] clarify paired cascade (#6386)

delan opened the pull request:
This patch makes edits for #6386.

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -667,12 +689,20 @@ Cascading and Per-Element Highlight Styles</h3>
+
   When any supported property is not given a value by the cascade,
-  it's value is determined by inheritance from
+  or given a value of ''inherit'',
```
Say “''inherit'' or ''unset''” here, and then delete the new paragraph below.

delan replied:
done in 06de02d057ea554fac09643037878d426b9941fb

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -603,10 +603,32 @@ Default UA Styles</h3>
-    :root::target-text    { color: MarkText; background: Mark; }
   </pre>
 
+  Note: the “paired cascade” behavior of [=highlight colors=]
```
We haven't introduced the paired cascading behavior yet, so moving the note that refers to it here is confusing.

delan replied:
yeah, it’s a combination of moving this note up from #highlight-cascade (because it’s about default UA styles) and moving the normative text for paired cascade down to #highlight-painting (because it affects used values), which leaves us with a problem here. i’ve tweaked the wording in fa44a161cdba4f8a4f117f53094478576d76ee32, what do you think?

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -603,10 +603,32 @@ Default UA Styles</h3>
+  for each highlight pseudo-element,
+  consisting of a 'color' and a 'background-color',
+  and doing so is recommended
+  for ''::selection'' and ''::target-text''.
```
Is this an RFC2119 RECOMMEND? Or did you mean something different?

delan replied:
4eb95b022c1eebb02bd4969bf723fd8ab4d6825a should make this relationship clearer.

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -758,6 +777,60 @@ Painting the Highlight</h3>
     </wpt>
 
+<h4 id="paired-cascade">
+Highlight Colors</h4>
```
I think this section can also be simplified a bit, but, I'll have to take a pass at it later. :)

delan replied:
by all means!

delan replied:
i’ve rewritten and heavily simplified this section in 86bb487445f17b354230b7832e116e697982d6de.

delan reviewed (commented):
Thanks for taking a look at this!

frivoal commented on css-pseudo-4/Overview.bs:
```diff
@@ -593,7 +601,7 @@ Styling Highlights</h3>
 
   Note: Historically (and at the time of writing)
-  only 'color' and 'background-color' have been interoperably supported.
+  only the highlight colors have been interoperably supported.
```
```suggestion
  only the [=highlight colors=] have been interoperably supported.
```

delan replied:
done in 24df479ce2bf25d3090559715a54e7eec31d4025

frivoal commented:
It seems to me that we might end up needing further tweaks based on https://github.com/w3c/csswg-drafts/issues/6818, but that other than that, this is probably good and merging it would be reasonable.

delan commented:
Thanks @frivoal! I think we can fairly cleanly leave #6818 for a separate patch, seeing as that issue is about resolved/used values and this patch only affects the cascade sections.

@fantasai did you want to take another pass at this before we merge?

delan commented:
@fantasai, @frivoal, some notable improvements to this patch since your last reviews:

* i’ve simplified the origins that affect paired cascade to author only (bfc6498f9fe6342474de16b936810b00e5b0e306)
* i’ve removed the text allowing UA tweaks to author highlight colors (0c46d9d9a15a65f9634cd4c72c6d5dbea27c8c12)
* i’ve defined the inherited value of ‘color’ in the root as ‘currentColor’ (ce778504e1adba403a311094fea9254378b4cf80)
* i’ve tied the default highlight colors for ::selection + ::target-text to system colors (4eb95b022c1eebb02bd4969bf723fd8ab4d6825a)
* i’ve rewritten and heavily simplified the new section about paired cascade itself (86bb487445f17b354230b7832e116e697982d6de)

I believe the new proposed text is clearer than ever, and conveys all of the details I wanted to convey (see pr description for the complete list) in far fewer words. Please take a look!

delan commented:
@fantasai, @frivoal, #6774 is now resolved, so these changes are ready for review. It looks like fantasai last reviewed 44aa6cede80d6a8fbda5e12b7e443e7c94c8407f while Florian last reviewed 24df479ce2bf25d3090559715a54e7eec31d4025, in case it helps you pick up where you left off and see what has changed.

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -724,10 +725,24 @@ Default UA Styles</h3>
-    :root::target-text    { color: MarkText; background: Mark; }
   </pre>
 
+  Some highlight pseudo-elements
+  should have <dfn>default highlight colors</dfn>,
+  consisting of a 'color' and a 'background-color'.
```
```suggestion
  Some [=highlight pseudo-elements=]
  have <dfn>paired default highlight colors</dfn>,
  consisting of a default 'color' and a 'background-color'
  that are either used or overridden together,
  see [[#paired-cascade]].
```

delan replied:
done in 4e092c568e8a4247309ffff98fae7b172f2b4067

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -724,10 +725,24 @@ Default UA Styles</h3>
+  Some highlight pseudo-elements
+  should have <dfn>default highlight colors</dfn>,
+  consisting of a 'color' and a 'background-color'.
+  For ''::selection'' they are
+  ''HighlightText'' and ''Highlight'',
```
```suggestion
  For ''::selection'' they should correspond to
  ''HighlightText'' and ''Highlight'',
```

delan replied:
done in 4e092c568e8a4247309ffff98fae7b172f2b4067

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -724,10 +725,24 @@ Default UA Styles</h3>
+  consisting of a 'color' and a 'background-color'.
+  For ''::selection'' they are
+  ''HighlightText'' and ''Highlight'',
+  while for ''::target-text'' they are
+  ''MarkText'' and ''Mark''.
```
Did we decide to add paired-cascading behavior to ''::target-text'' also? There's afaik no Web-compatibility requirement for it here, right?

delan replied:
not formally, and not for web compat, just based on impl opinions from Emilio and i (<https://github.com/w3c/csswg-drafts/issues/6386#issuecomment-957389344>).

delan replied:
s/are/should correspond to/ done in https://github.com/w3c/csswg-drafts/commit/4e092c568e8a4247309ffff98fae7b172f2b4067

fantasai replied:
> just based on impl opinions from Emilio and i (https://github.com/w3c/csswg-drafts/issues/6386#issuecomment-957389344).

I think this deserves its own issue and WG resolution

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -724,10 +725,24 @@ Default UA Styles</h3>
+  while for ''::target-text'' they are
+  ''MarkText'' and ''Mark''.
+
+  Note: the painting behavior of highlight colors
+  means that default highlight colors
+  can't be expressed in terms of 'color' and 'background-color'
+  in [=UA style sheet=] rules.
```
Can you explain why this wording is better than the original wording?

From the author's perspective (though you might implement it however you want), it's a special cascading rule, not a “painting” behavior, so I think the original wording is more understandable.

Also, you want `'color' and 'background-color' declarations` here.

delan replied:
i had to reword it because i removed the preceding sentence

> The UA must use its own highlight colors for [::selection](https://drafts.csswg.org/css-pseudo-4/#selectordef-selection) only when neither [color](https://drafts.csswg.org/css-color-4/#propdef-color) nor [background-color](https://drafts.csswg.org/css-backgrounds-3/#propdef-background-color) has been specified by the author.

which in turn was removed because it’s now explained more thoroughly (and correctly) in #highlight-painting, but i see where you’re coming from. i’ll revert to the old text and correct the thing about ::selection.

delan replied:
done in 26b6c2b32e30d2ef84cc48365939ce682eb17482

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -879,6 +890,43 @@ Painting the Highlight</h3>
     css/css-pseudo/highlight-painting-003.html
     </wpt>
 
+<h4 id="paired-cascade">
+Highlight Colors</h4>
```
This should be a subsection of “Default UA Styles”.

delan replied:
i think you’re right, but only if we make it clearer for implementors that this is a used-value-time thing that doesn’t affect the computed value. imo it would be enough by linkifying “used” to used value again and not using the term “paired cascade”, despite its historical relevance as what we used in the note and in #6386

delan replied:
done in e94418630480cf5c9ac640b207e446b33e890cc7

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -879,6 +890,43 @@ Painting the Highlight</h3>
+  must only be [=used value|used=]
+  when neither of the [=highlight colors=]
+  yield a [=cascaded value=]
+  or [=specified value=]
```
iirc i was trying to say that paired cascade’s suppression of defaults when authors set highlight colors spreads to descendants. for example, if the author only declares body::selection { color: red; }, then we want p::selection to be red on initial (transparent), not red on default (Highlight), even though there is no author cascaded value, only an author specified value after (implicit) defaulting.

same goes for if the UA or user also declares p::selection { color: inherit; }. in this case, there is no author cascaded value, but there *is* an author specified value after (explicit) defaulting.

hopefully that makes sense, my mental model of css-cascade is pretty new and was only based on reading the spec. either way, i agree that those three words don’t really convey this idea well.

delan replied:
done in e554b21821e32c57d746e7a70117b5c1aa4848db

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -879,6 +890,43 @@ Painting the Highlight</h3>
+      em::selection { color: unset; }
+      em::selection { color: green; }
+
+      /* specified value is author origin */
```
same thing as my big comment about specified value. the first four rules are the author directly setting ‘color’ on the em, so default highlight colors are suppressed due to author cascaded value, while the last rule is the author setting ‘color’ on the parent, so there is no author cascaded value, but they are suppressed due to author specified value.

these comments were only meant to make up for the lack of clarity there, so i’ll remove them.

delan replied:
done in 4e092c568e8a4247309ffff98fae7b172f2b4067

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -696,11 +696,12 @@ Styling Highlights</h3>
 
   Issue: Are there any other properties that should be included here?
 
-  The 'color' property specifies the color of both the text
-  and all line decorations (underline, overline, line-through)
-  and emphasis marks ('text-emphasis')
-  applied to the text
-  by the <a>originating element</a> and its ancestors and descendants.
```
we already explain this in #highlight-painting (and mention emphasis marks in a note):

> A highlight pseudo-element suppresses the normal drawing of any associated text, and the text decorations (other than shadows) that had been applied to that text. Instead the topmost active highlight overlay redraws that text (and those decorations) over all the highlight overlay backgrounds using that highlight’s own [color](https://drafts.csswg.org/css-color-4/#propdef-color).

i don’t mind redundancy, but the way i interpreted the layout of the highlight pseudos section is that computed-value-time behaviour belonged in #highlight-styling, while used-value-time behaviour belonged in #highlight-painting.

delan replied:
(that said, you’ve made a good point elsewhere about that distinction not being important to authors)

fantasai replied:
OK, in that case maybe make it a Note? That way the information is presented simply here, but the normative text is elsewhere. (You do that by just adding "Note: " at the start of the paragraph.)

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -696,11 +696,12 @@ Styling Highlights</h3>
-  by the <a>originating element</a> and its ancestors and descendants.
+  'color' and 'background-color'
+  are known as <dfn>highlight colors</dfn>.
+  Because this concept exists for compatibility reasons,
+  other similar properties
+  like 'fill-color' or 'stroke-color'
+  must not be treated as highlight colors.
```
it looks like i moved this definition up in 65597ae6af59211f1c857c6509cf696bcdbf52df. at the time it linked to the section explaining the impl impact, and i wanted to use the term “highlight colors” in the “Historically” note, but neither of those are true now, so i’ll move it back.

delan replied:
done in 6a142341ace64adff79c1f622025630fb131b67a. i also eliminated the term “highlight colors” completely, which existed in the current draft but doesn’t really add any value as a separate term to “paired default highlight colors”, since the set of properties will never change (and we’re even adding a note to clarify this).

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -696,6 +696,14 @@ Styling Highlights</h3>
+  Because this concept exists for compatibility reasons,
+  other similar properties
+  like 'fill-color'
+  must not be treated as highlight colors.
```
done in 4e092c568e8a4247309ffff98fae7b172f2b4067

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -696,11 +696,12 @@ Styling Highlights</h3>
-  applied to the text
-  by the <a>originating element</a> and its ancestors and descendants.
+  'color' and 'background-color'
+  are known as <dfn>highlight colors</dfn>.
```
done in 4e092c568e8a4247309ffff98fae7b172f2b4067

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -714,7 +715,7 @@ Styling Highlights</h3>
 
   Note: Historically (and at the time of writing)
-  only 'color' and 'background-color' have been interoperably supported.
+  only the highlight colors have been interoperably supported.
```
done in 4e092c568e8a4247309ffff98fae7b172f2b4067

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -788,12 +803,19 @@ Cascading and Per-Element Highlight Styles</h3>
   of its <a>originating element</a>'s parent element
   (regardless of whether that property is an <a>inherited property</a>).
 
+  For the root element,
```
done in 4e092c568e8a4247309ffff98fae7b172f2b4067

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -879,6 +890,43 @@ Painting the Highlight</h3>
+
+  For compatibility reasons,
+  [=default highlight colors=]
+  must only be [=used value|used=]
```
done in 4e092c568e8a4247309ffff98fae7b172f2b4067

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -879,6 +890,43 @@ Painting the Highlight</h3>
+  of the [=author origin=].
+  When a highlight color is ''revert'' or ''revert-layer'',
+  the origin <em>after</em> rolling back the cascade
+  determines whether this is the case.
```
done in 4e092c568e8a4247309ffff98fae7b172f2b4067

delan commented on css-pseudo-4/Overview.bs:
```diff
@@ -724,10 +732,33 @@ Default UA Styles</h3>
+  for each highlight pseudo-element,
+  consisting of a 'color' and a 'background-color',
+  and doing so is recommended
+  for ''::selection'' and ''::target-text''.
```
i think these tweaks are already true in the current version?

fantasai commented on css-pseudo-4/Overview.bs:
```diff
@@ -724,16 +718,68 @@ Default UA Styles</h3>
   </pre>
 
+  Some [=highlight pseudo-elements=]
+  should have <dfn>paired default highlight colors</dfn>,
+  consisting of a default 'color' and 'background-color'
```
```suggestion
  should have <dfn>paired default highlight colors</dfn>--
  a default 'color' and 'background-color'
```

delan replied:
done in 5084397ba8f811b011f8e6680489e7dff884298b

fantasai commented:
Ok, looks good. Please SQUASH and rebase before merging. :)

delan commented:
Thank you both for all of your feedback!
